profile-widget/profilewidget2.cpp

```diff
@@ -1974,26 +1977,46 @@ void ProfileWidget2::keyEscAction()
 		plannerModel->cancelPlan();
 }
 
-void ProfileWidget2::plotPictures()
+void ProfileWidget2::clearPictures()
 {
-	Q_FOREACH (DivePictureItem *item, pictures) {
-		item->hide();
-		item->deleteLater();
-	}
 	pictures.clear();
+}
 
-	if (printMode)
-		return;
+void ProfileWidget2::updatePictures(const QModelIndex &from, const QModelIndex &to)
+{
+	DivePictureModel *m = DivePictureModel::instance();
+	for (int picNr = from.row(); picNr <= to.row(); ++picNr) {
+		int picItemNr = picNr - m->rowDDStart;
+		if (picItemNr < 0 || (size_t)picItemNr >= pictures.size())
+			return;
+		if (!pictures[picItemNr])
+			return;
 
-	double x, y, lastX = -1.0, lastY = -1.0;
+		pictures[picItemNr]->setPixmap(m->index(picNr, 0).data(Qt::UserRole).value<QPixmap>());
+	}
+}
+
+void ProfileWidget2::plotPictures()
+{
 	DivePictureModel *m = DivePictureModel::instance();
+	pictures.resize(m->rowDDEnd - m->rowDDStart);
+
+	double x, y, lastX = -1.0, lastY = -1.0;
 	for (int i = m->rowDDStart; i < m->rowDDEnd; i++) {
+		int picItemNr = i - m->rowDDStart;
 		int offsetSeconds = m->index(i, 1).data(Qt::UserRole).value<int>();
 		// it's a correct picture, but doesn't have a timestamp: only show on the widget near the
-		// information area.
-		if (!offsetSeconds)
+		// information area. A null pointer in the pictures array indicates that this picture is not
+		// shown.
+		if (!offsetSeconds) {
+			pictures[picItemNr].reset();
 			continue;
-		DivePictureItem *item = new DivePictureItem();
+		}
+		if (!pictures[picItemNr]) {
+			pictures[picItemNr].reset(new DivePictureItem);
+			scene()->addItem(pictures[picItemNr].get());
+		}
+		DivePictureItem *item = pictures[picItemNr].get();
```

Comment:
Suggestion: if you move this line before the `if`, you can save yourself from repeating `pictures[picItemNr]` three times. It's ok to `.get()` a null unique pointer -- you get nullptr.